### PR TITLE
Implementation of stylist ui

### PR DESCRIPTION
**Pull request by Asheraf**
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

[//]: # (Note: Lines beginning with syntax such as this one, are comments and will not be visible in your report!)

### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

this adds support for stylist ui in 2015+ clients

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/2004)

File: src/map/clif.c [hunk 1]
```diff
@@ -20226,6 +20226,185 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+	VECTOR_ENSURE(stylist_data[type], 1, 1);
+	VECTOR_PUSH(stylist_data[type], entry);
+	return true;
+}
+
+bool clif_style_change_validate_requirements(struct map_session_data *sd, int type, uint16 idx)
+{
+	struct item it;
+	struct stylist_data_entry *entry;
+
```
Comment (4144) [inline src/map/clif.c]: missing assert or check for idx

File: src/map/clif.c [hunk 2]
```diff
@@ -20226,6 +20226,185 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+	nullpo_retv(sd);
+
+	msg.receiver_id = sd->status.char_id;
+	msg.items[0].item.nameid = itemid;
+	msg.items[0].item.amount = 1;
+	msg.items[0].item.identify = 1;
+	msg.type = MAIL_TYPE_NPC | MAIL_TYPE_ITEM;
+
+	safestrncpy(msg.sender_name, "Styling Shop", NAME_LENGTH);
```
Comment (4144) [inline src/map/clif.c]: probably this strings better read from messages.conf?

File: src/map/clif.c [hunk 3]
```diff
@@ -20226,6 +20226,187 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+bool clif_stylist_read_db_libconfig_sub(struct config_setting_t *it, int idx, const char *source)
+{
+	struct stylist_data_entry entry = { 0 };
+	int i32 = 0, type = 0;
+	int64 i64 = 0;
+
+	nullpo_ret(it);
+	nullpo_ret(source);
+
+	if (!itemdb->lookup_const(it, "Type", &type) || type > MAX_STYLIST_TYPE || type <= 0) {
```
Comment (4144) [inline src/map/clif.c]: should be ``type > MAX_STYLIST_TYPE`` because if type == MAX_STYLIST_TYPE this is invalid index.

File: src/map/clif.c [hunk 4]
```diff
@@ -20226,6 +20226,187 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+	return true;
+}
+
+bool clif_style_change_validate_requirements(struct map_session_data *sd, int type, uint16 idx)
+{
+	struct item it;
+	struct stylist_data_entry *entry;
+
+	nullpo_retr(false, sd);
+	Assert_retr(false, type > 0 && type <= MAX_STYLIST_TYPE);
```
Comment (4144) [inline src/map/clif.c]: should be ``type >= 0 && type < MAX_STYLIST_TYPE``

File: src/map/clif.c [hunk 5]
```diff
@@ -20226,6 +20226,187 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+		if (clif->style_change_validate_requirements(sd, LOOK_HEAD_BOTTOM, p->BottomAccessory - 1))
+			clif->stylist_send_rodexitem(sd, entry->id);
+	}
+	clif->style_change_response(sd, true);
+	return;
+}
+
+void clif_style_change_response(struct map_session_data *sd, int8 flag)
+{
+	struct PACKET_ZC_STYLE_CHANGE_RES p;
```
Comment (4144) [inline src/map/clif.c]: missing nullpo check for sd

File: src/map/clif.c [hunk 6]
```diff
@@ -20226,6 +20226,189 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+}
+
+bool clif_style_change_validate_requirements(struct map_session_data *sd, int type, uint16 idx)
+{
+	struct item it;
+	struct stylist_data_entry *entry;
+
+	nullpo_retr(false, sd);
+	Assert_retr(false, type >= 0 && type < MAX_STYLIST_TYPE);
+	Assert_retr(false, idx <= VECTOR_LENGTH(stylist_data[type]));
```
Comment (4144) [inline src/map/clif.c]: should be ``idx < VECTOR_LENGTH(stylist_data[type]))``

Comment (Asheraf) [reply src/map/clif.c]: `VECTOR_LENGTH(stylist_data[type]))` is valid index in here

Comment (4144) [reply src/map/clif.c]: idx used as vector index. index must be from 0 to len-1. because if use index with len, it will be one item outside of array.

Comment (Asheraf) [reply src/map/clif.c]: ah yes my mistake, i was using the index from the packet directly which starts from `1 to len`, fixed.

File: src/map/script.c [hunk 7]
```diff
@@ -24000,6 +24000,16 @@ BUILDIN(hateffect)
...
 }
 
+BUILDIN(openstylist)
+{
+	struct map_session_data *sd = script_rid2sd(st);
+
+	if (sd == NULL)
+		return false;
+
+	clif->open_ui(sd, STYLIST_UI);
```
Comment (4144) [inline src/map/script.c]: missing return true; here

Comment (4144) [reply src/map/script.c]: look like because this travis failed

File: src/map/clif.c [hunk 8]
```diff
@@ -20601,6 +20605,193 @@ void clif_ui_action(struct map_session_data *sd, int32 UIType, int32 data)
...
+	safestrncpy(msg.sender_name, msg_txt(366), NAME_LENGTH);
+	safestrncpy(msg.title, msg_txt(367), RODEX_TITLE_LENGTH);
+	safestrncpy(msg.body, msg_txt(368), MAIL_BODY_LENGTH);
+	msg.send_date = (int)time(NULL);
+	msg.expire_date = (int)time(NULL) + RODEX_EXPIRE;
+
+	intif->rodex_sendmail(&msg);
+}
+
+void clif_parse_cz_req_style_change(int fd, struct map_session_data *sd)
```
Comment (4144) [inline src/map/clif.c]: missing nonnull attributes same like in other clif_parse_* functions.
Also next nullpo_retv(sd) useless if add nonnull attributes for sd parameter

**Comment by MishimaHaruna** [comment]
@Asheraf there's a minor conflict in script_commands.txt that needs to be resolved before this is merged. Everything else should be ready unless @4144 has something to say about the fixes to his requested changes

**Comment by Asheraf** [comment]
@MishimaHaruna resolved.

File: src/map/clif.c [hunk 9]
```diff
@@ -20672,6 +20676,192 @@ void clif_private_airship_response(struct map_session_data *sd, uint32 flag)
...
+				pc->changelook(sd, type, entry->id);
+			else
+				clif->stylist_send_rodexitem(sd, entry->id);
+		}
+	}
+}
+
+void clif_style_change_response(struct map_session_data *sd, int8 flag)
+{
+	struct PACKET_ZC_STYLE_CHANGE_RES p;
```
Comment (4144) [inline src/map/clif.c]: here need add limit for send packet to only supported clients
``#if PACKETVER >= 20151104``

**Pull request merged**